Adding shared template option

igor-ribeiiro opened the pull request:
Thank you for your pull request. Please provide the following information.

---

1. **Do these changes impact *User Experience*?** (e.g., how the user interacts with Yo Office and/or the files and folders the user sees in the project that Yo Office creates)
    > 
    > * [X]  Yes
    > * [ ]  No

There will be a new option for the shared runtime template.

2. **Do these changes impact *documentation*?** (e.g., a tutorial on https://docs.microsoft.com/en-us/office/dev/add-ins/overview/office-add-ins)
    > 
    > * [X]  Yes
    > * [ ]  No

The tutorial on how to create CFs need to be updated. It also needs to include shared vs legacy runtimes differences.


3. **Validation/testing performed**:

Tested creating a new yo office app and choosing the option shared template. The template is indeed the shared one.

4. **Platforms tested**:

    > * [X] Windows
    > * [ ] Mac

igor-ribeiiro commented on src/app/config/projectProperties.json:
```diff
@@ -129,6 +129,25 @@
             ]
         },
+        "excel-functions-shared": {
+            "displayname": "Excel Shared Runtime Custom Functions Add-in project",
```
I don't think there's a name yet. I will keep this PR on hold until we get the name confirmation and then change when we have it

Rick-Kirkham replied:
@alison-mk can help choose a name. If it were me, I'd like to avoid so many stacked nouns. I'd prefer breaking it up with a prepositional phrase. "Excel Custom Functions with a Shared Runtime project".

akrantz replied:
"Excel Custom Functions using Shared Runtime"  ?

igor-ribeiiro replied:
We can also change the normal Excel CFs name. Please let me know if we need to do that.

alison-mk replied:
I like Ricky and Adam's suggestions. Could we also place Shared Runtime before JS-only in the list, to subtly indicate that it's the preferred method? @igor-ribeiiro Yes, I would also like to change the existing CF template name. I recommend: 

- Excel Custom Functions using a Shared Runtime
- Excel Custom Functions using a JavaScript-only Runtime

igor-ribeiiro commented:
Added the shared option before the manifest option and reordered the options for most used to least used.
The change https://github.com/OfficeDev/generator-office/pull/714/files caused a bug which the message wouldn't have the host in it:
![image](https://user-images.githubusercontent.com/18221895/174691001-4e190008-57c5-43b8-b907-f653ff1eca27.png)

So I also fixed that bug.

millerds commented on src/app/index.ts:
```diff
@@ -196,6 +196,9 @@ module.exports = class extends yo {
         }
       ];
       const answerForScriptType = await this.prompt(askForScriptType);
+      if (!answerForScriptType.scriptType && !this.options.ts) {
```
We wanted to leave these off because the default should be set by the question settings.

millerds replied:
I'm not seeing that in the documentation and if it really didn't have the default answer then wouldn't there be an error when trying to create the project since there wouldn't be an answer (i.e. something like a null exception)?

millerds commented on src/app/index.ts:
```diff
@@ -425,10 +426,12 @@ module.exports = class extends yo {
     this.log(`  ${chalk.bgGreen('projectType')}:Specifies the type of project to create. Valid project types include:`);
     this.log(`    ${chalk.yellow('angular:')}  Creates an Office add-in using Angular framework.`);
-    this.log(`    ${chalk.yellow('excel-functions:')} Creates an Office add-in for Excel custom functions.  Must specify 'Excel' as host parameter.`);
+    this.log(`    ${chalk.yellow('excel-functions-shared:')} Creates an Office add-in for Excel custom functions using a Shared Runtime.  Must specify 'Excel' as host parameter.`);
```
Don't we automatically set the host parameter?  We wouldn't need this part of the message.

igor-ribeiiro replied:
I guess it's because the messages are a bit different between themselves. We would need to add a new parameter to all hosts (like `displayMessage`) if we would like to keep current messages.

millerds commented on src/app/index.ts:
```diff
@@ -425,10 +426,12 @@ module.exports = class extends yo {
     this.log(`    ${chalk.yellow('jquery:')} Creates an Office add-in using Jquery framework.`);
     this.log(`    ${chalk.yellow('manifest:')} Creates an only the manifest file for an Office add-in.`);
     this.log(`    ${chalk.yellow('react:')} Creates an Office add-in using React framework.\n`);
+    this.log(`    ${chalk.yellow('teams-manifest:')} Creates Outlook Add-in with Teams Manifest (Developer preview).\n`);
```
Why is this listed as an add?  Shouldn't it already be there?

igor-ribeiiro commented:
@millerds "I'm not seeing that in the documentation and if it really didn't have the default answer then wouldn't there be an error when trying to create the project since there wouldn't be an answer (i.e. something like a null exception)?"

It is null that gets returned (or something close to it). We don't get an error because we check if it exists:
```
        scriptType: answerForScriptType.scriptType ? answerForScriptType.scriptType : this.options.ts ? typescript : javascript
```

The `default` is only the value that will get chosen when you press "enter", without choosing a particular option.

millerds commented:
So if that line is taking care of the question not asked case, then there is no need for that if block.  It looks like the bug you are saying you are fixing is not with the language chosen, but rather the host (that is what is missing from the message).  The host also has a fall back in the same place as the language, but that isn't using the ?: case but rather the || syntax and probably behaves differently.

The point is that we shouldn't need two places to fall back onto and I prefer the _configureProject method with a simple value precedent check rather than having to maintain multiple conditionals with the same tests (the one for weather or not the question should be asked, and the same one for weather or not the question was asked.

igor-ribeiiro commented:
> Member

Sure, we can handle all these cases handling at the same place. I made some changes so that all cases are covered in a single line and hopefully it is not too confusing.

millerds commented on src/app/index.ts:
```diff
@@ -295,12 +290,12 @@ module.exports = class extends yo {
-        scriptType: answerForScriptType.scriptType ? answerForScriptType.scriptType : this.options.ts ? typescript : javascript
       };
+      this.project.host = answerForHost.host ? answerForHost.host : this.options.host ? this.options.host : jsonData.getHostTemplateNames(this.project.projectType)[0];
+      this.project.scriptType = answerForScriptType.scriptType ? answerForScriptType.scriptType : this.options.ts ? typescript : this.options.js ? this.options.js : jsonData.getSupportedScriptTypes(this.project.projectType)[0];
```
I think the "this.options.js ? this.options.js" is intended to be "this.options.js ? javascript"

Could also use some formatting to encourage better readability rather than all on one line.

millerds commented on src/app/index.ts:
```diff
@@ -29,6 +29,7 @@ let language;
 const manifest = 'manifest';
 const sso = 'single-sign-on';
 const typescript = `TypeScript`;
+let jsonData;
```
Why pull this out of the try block?

igor-ribeiiro replied:
It is being used on another function. So I wanted to let this other function `configureProject` also have access to `jsonData`

millerds commented on src/app/index.ts:
```diff
@@ -295,12 +290,12 @@ module.exports = class extends yo {
         isExcelFunctionsProject: isExcelFunctionsProject,
-        scriptType: answerForScriptType.scriptType ? answerForScriptType.scriptType : this.options.ts ? typescript : javascript
       };
+      this.project.host = answerForHost.host ? answerForHost.host : this.options.host ? this.options.host : jsonData.getHostTemplateNames(this.project.projectType)[0];
```
Why is this moved out the block it was in?

millerds commented on src/app/index.ts:
```diff
@@ -289,12 +290,17 @@ module.exports = class extends yo {
         isExcelFunctionsProject: isExcelFunctionsProject,
-        scriptType: answerForScriptType.scriptType ? answerForScriptType.scriptType : this.options.ts ? typescript : javascript
       };
+      this.project.host = answerForHost.host ? answerForHost.host 
```
Why not leave this in the block using "host: answerforHost.host . . ." with everything else (you don't need to use "this.project.")?

millerds replied:
They set projectType inside the block as well.  It's a big deal but I would add 'const projType = _.toLower(this.options.projectType) || _.toLower(answerForProjectType.projectType)' before the block then create everything inside the block using the projType variable.

millerds commented on src/app/index.ts:
```diff
@@ -289,12 +290,17 @@ module.exports = class extends yo {
       };
+      this.project.host = answerForHost.host ? answerForHost.host 
+        : this.options.host ? this.options.host 
+        : jsonData.getHostTemplateNames(this.project.projectType)[0];
```
If this is coming from outside of this method then we should check that it's not empty.

millerds commented on src/app/index.ts:
```diff
@@ -287,20 +287,35 @@ module.exports = class extends yo {
+        host: answerForHost.host
+          ? answerForHost.host
+          : this.options.host
+          ? this.options.host
```
Add a new indent level here.

igor-ribeiiro replied:
Using vscode format selection

millerds commented on src/app/index.ts:
```diff
@@ -287,20 +287,35 @@ module.exports = class extends yo {
+        scriptType: answerForScriptType.scriptType
+          ? answerForScriptType.scriptType
+          : this.options.ts
+          ? typescript
```
Add an indent level here.

igor-ribeiiro replied:
Same as other comment, I'm using `format selection` from vscode

millerds commented on src/app/index.ts:
```diff
@@ -287,20 +287,35 @@ module.exports = class extends yo {
+          : this.options.ts
+          ? typescript
+          : this.options.js
+          ? javascript
```
Add an indent level here.

igor-ribeiiro replied:
I'm using the prettier here. So those two spaces are the recommended amount.

igor-ribeiiro commented on src/app/index.ts:
```diff
@@ -287,20 +287,35 @@ module.exports = class extends yo {
-        : this.options.js ? javascript
-        : jsonData.getSupportedScriptTypes(this.project.projectType)[0];
+
+      if (!this.project.host) {
```
It should never be null unless there is an error with our code. I will just check for `jsonData` null

igor-ribeiiro commented on src/app/index.ts:
```diff
@@ -295,7 +295,7 @@ module.exports = class extends yo {
           : this.options.host
           ? this.options.host
-          : jsonData.getHostTemplateNames(projType)[0],
+          : jsonData?.getHostTemplateNames(projType)[0],
```
It would get host = `null` which would error out later on the code in the `npm run convertToSingleHost` line. 
JSON data is currently never null. If it were to be null, then there is an with our code itself not related to a client's path of choices.

In that case, do you think throwing an error if it is null would help? Ideally, this error never appears, but it might happen if we make mistakes in our code in the future.

millerds replied:
It's probably OK as is.  You are right in that it's a safety for us rather than customers.  I just wanted to make sure there wouldn't be some other cryptic error or silent failure if null was encountered.

igor-ribeiiro commented:
@millerds could you review it again? Needed to comment some SSO tests that were added back during the merge
